### PR TITLE
Implement OpenSSL secure memory for Windows

### PR DESCRIPTION
**Pull request by jgowdy**
Hello, please review this pull request implementing CRYPTO_secure_malloc and the associated functions for the Windows platform using equivalent functions such as VirtualAlloc for mmap, etc.  Thank you.

##### Checklist

Documentation for CRYPTO_secure_malloc(3) doesn't seem to mention which platforms are and are not supported, so I do not believe any update is required.

Existing secmemtest should cover Windows implementation

### CODE REVIEW

File: crypto/mem_sec.c
Comment (paulidale) [inline crypto/mem_sec.c:30]: NIT: indentation

File: crypto/mem_sec.c
Comment (paulidale) [inline crypto/mem_sec.c:34]: NIT: indentation

File: e_os.h
Comment (paulidale) [inline e_os.h:365]: Break the `|| defined(_WIN32)` onto the next line.  The indentation is also off with the addition of the initial `(`.

File: crypto/mem_sec.c
Comment (paulidale) [inline crypto/mem_sec.c:468]: There seems to be a variety of `defined(_WIN32)` and `!defined(_WIN32)`.  I don't know if making them consistent would be better or not.  Thoughts anyone?

**Review by paulidale — COMMENTED** [review]
This looks good, I'm not Windows savvy though.  Just a few [style](https://www.openssl.org/policies/codingstyle.html) nitpicks.

**Comment by jgowdy** [comment]
Great, happy to fix up those nits!  Any other feedback is welcome.

File: crypto/mem_sec.c
Comment (mattcaswell) [inline crypto/mem_sec.c:492]: According to this page:

https://docs.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc

"Memory allocated by this function is automatically initialized to zero."

So this memset seems redundant?

File: crypto/mem_sec.c
Comment (mattcaswell) [inline crypto/mem_sec.c:531]: It seems there could be problems here. According to the VirtuaLock documentation:

"The maximum number of pages that a process can lock is equal to the number of pages in its minimum working set minus a small overhead."

And according to the SetProcessWorkingSetSize documentation for `dwMinimumWorkingSetSize`:

"The default size is 50 pages (for example, this is 204,800 bytes on systems with a 4K page size)."

So, it seems a typical value for the maximum amount of secure memory in a process might be 200k. I have no idea how much secure memory a typical process uses, but I wonder if it ever approaches 200k? Perhaps not, but I wonder if we should document this limitation somewhere (probably on the `CRYPTO_secure_malloc_init` man page) and how to increase this?

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I'm not really a Windows guy either, but from my reading of the API docs this looks good aside from the couple of comments below.

**Comment by jgowdy** [comment]
You're right, I'll take that out.

File: crypto/mem_sec.c
Comment (paulidale) [reply crypto/mem_sec.c:531]: @richsalz, any idea about the sizing here?

Linux/Unix isn't going to permit unlimited size either.  `ulimit -l` can restrict it further.

Comment (richsalz) [reply crypto/mem_sec.c:531]: You'll have to do some estimating, based on the number of private keys and the number of simultaneous sessions you expect to have.  There is no good answer.  Perhaps adding a couple of sentences like that to the manpage would be useful.  The size we use is far above what would be typical for a common origin 😃

Comment (jgowdy) [reply crypto/mem_sec.c:531]: Yeah, as someone who consumes OpenSSL's secure heap from a managed language, on Linux the default is 64KB, so the situation is actually much worse on other platforms.

File: crypto/mem_sec.c
Comment (jgowdy) [reply crypto/mem_sec.c:492]: You're right, I'll remove this.

File: crypto/mem_sec.c
Comment (jgowdy) [reply crypto/mem_sec.c:529]: Given that the default limit for Windows is larger than the limit for Linux, I'm inclined to consider this a non-issue, but I'd love more feedback.  We can certainly adjust the minimum and maximum working set size if we think that makes sense, but we'd likely want to do the same setrlimit call on other platforms too.  Personally given the comparison to Linux defaults, I'd make introducing limit adjustments to be another effort.

File: e_os.h
Comment (jgowdy) [reply e_os.h]: I'm hoping I got this right now, can you double check me please?

**Comment by jgowdy** [comment]
If you don't mind, could you possibly label this PR as 'hacktoberfest-accepted'?  I did *not* do this pull request for the purpose of hacktoberfest, but I am participating.  Thanks!

**Comment by mattcaswell** [comment]
> If you don't mind, could you possibly label this PR as 'hacktoberfest-accepted'? 

Could you explain what the label actually means? I'm not familiar with it.

**Comment by paulidale** [comment]
And how to add it...

**Comment by jgowdy** [comment]
It's just a label on the PR that allows the PR to count towards the goals of https://hacktoberfest.digitalocean.com which is a challenge for people to contribute towards open source projects in October.

It's not a big deal, but if you would set a label field on the PR to that string I would get credit for it.  

The itself PR is what matters though.  I didn't make the PR for the sake of Hacktoberfest.

**Comment by t8m** [comment]
I am not sure we want to add such label to our already quite long list of labels. They suggest that making a PR on your own repository fork will be also counted.

**Comment by mspncp** [comment]
According to https://hacktoberfest.digitalocean.com

image

it needs to be a [GitHub topic](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/classifying-your-repository-with-topics) not a label.

**Comment by mspncp** [comment]
> According to https://hacktoberfest.digitalocean.com
> 
> image
> 
> it needs to be a [GitHub topic](https://docs.github.com/en/free-pro-team@latest/github/administering-a-repository/classifying-your-repository-with-topics) not a label.

C'mon @openssl/omc, why not add it? It might attract lotsof pullrequests ;-)

**Comment by mspncp** [comment]
@openssl/omc talking about topics: maybe we should also consider other more serious topics to our repo?

**Comment by mspncp** [comment]
> > If you don't mind, could you possibly label this PR as 'hacktoberfest-accepted'?
> 
> Could you explain what the label actually means? I'm not familiar with it.

Ok, so there is also a label for explicit opt-in:  https://hacktoberfest.digitalocean.com/hacktoberfest-update

File: crypto/mem_sec.c
Comment (mattcaswell) [reply crypto/mem_sec.c:529]: Ok. Let's leave it at that for this PR. Wouldn't it be nice if the arena size could dynamically expand and contract as required? If we had that feature then this could probably be configured by default through the auto initialisation rather than requiring explicit init and then all users could benefit from it.

**Review by mattcaswell — APPROVED** [review]
LGTM. Needs second review.

**Comment by mattcaswell** [comment]
I'm ok with that.

**Comment by paulidale** [comment]
I'm fine with the label for this or for the project to participate (although I worry about attracting lots of extra work).

**Comment by mspncp** [comment]
Don’t worry @paulidale, only ten days to go ;-)

**Review by paulidale — APPROVED** [review]

**Comment by paulidale** [comment]
I've posited the hacktoberfest question to the project list.

**Comment by mspncp** [comment]
> I've posited the [hacktoberfest question](https://www.mail-archive.com/openssl-project@openssl.org/msg02208.html) to the project list.

@jgowdy thank you for your contribution! Your pull request has reached the required number of approvals and will be merged after the 24 hour grace period. @mattcaswell and @paulidale already indicated consent to your request for a Hacktoberfest badge,  but I'll wait another day or so to add it, in order not to preempt the discussion on the mailing list.

**Comment by jgowdy** [comment]
Excellent, thank you all!

**Comment by jgowdy** [comment]
I've actually completed the Hacktoberfest challenge already, so we can skip the label.  I don't want to pollute the label list for no reason.  I appreciate the consideration.

**Comment by paulidale** [comment]
Merged to master.  Thanks.

**Pull request closed without merging**
